### PR TITLE
Propagate theme attribute

### PR DESCRIPTION
**Pull request by yuriy-fix**
Propagate theme attribute to nested overlays, list-boxes and items.

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-context-menu/238)

### CODE REVIEW

File: src/vaadin-context-menu-overlay.html [hunk 1]
```diff
@@ -69,6 +69,12 @@
             parentOverlay: {
               type: Object,
               readOnly: true
+            },
+
+            theme: {
```
Comment (tomivirkki) [inline src/vaadin-context-menu-overlay.html]: This shouldn't be added. It's already defined in theme property mixin

Comment (yuriy-fix) [reply src/vaadin-context-menu-overlay.html]: Removed.

Comment (tomivirkki) [reply src/vaadin-context-menu-overlay.html]: No observer was added, what invokes `_themeChanged` now?

File: src/vaadin-contextmenu-items-mixin.html [hunk 2]
```diff
@@ -236,7 +243,7 @@
     __initMenu(root, menu) {
       if (!root.firstElementChild) {
         root.innerHTML = `
-          <vaadin-context-menu-list-box></vaadin-context-menu-list-box>
+          <vaadin-context-menu-list-box theme="${this.theme}"></vaadin-context-menu-list-box>
```
Comment (tomivirkki) [inline src/vaadin-contextmenu-items-mixin.html]: Make sure the theme attribute isn't "null" or "undefined" or event "" at any time.

File: src/vaadin-context-menu-overlay.html [hunk 3]
```diff
@@ -90,6 +94,31 @@
           });
         }
 
+        _themeChanged(theme) {
+          const childContextMenu = this.querySelector('vaadin-context-menu');
```
Comment (tomivirkki) [inline src/vaadin-context-menu-overlay.html]: Remove this logic. Theme changes, while the overlay is open, are very unlikely. To be on the safe side, we could close the overlay & empty the root element here instead.

File: src/vaadin-contextmenu-items-mixin.html [hunk 4]
```diff
@@ -156,6 +156,12 @@
       // Store the reference to align based on parent overlay coordinates
       subMenu.$.overlay._setParentOverlay(parent);
 
+      // Set theme attribute from parent element
+      const parentTheme = parent.getAttribute('theme');
+      if (parentTheme && parentTheme !== '') {
```
Comment (tomivirkki) [inline src/vaadin-contextmenu-items-mixin.html]: `if (parentTheme)` should cover both cases ('' is also falsy)

File: test/items.html [hunk 5]
```diff
@@ -414,6 +422,97 @@
...
+            subMenu = getSubMenu();
+            flush(() => {
+              open(menuComponents(subMenu)[1]);
+              subMenu2 = getSubMenu();
+              flush(done);
+            });
+          });
+        });
+
+        function themeAttributeEquals(theme, el, overlay = false, listBox = false, items = false) {
```
Comment (tomivirkki) [inline test/items.html]: This helper is strange. It runs different checks depending on values of arbitrary boolean parameters that we then enable/disable differently for different tests...just test all the attributes in one go.

File: test/items.html [hunk 6]
```diff
@@ -414,6 +422,97 @@
...
+          } else if (listBox) {
+            return elListBox.getAttribute('theme') === theme;
+          } else if (items) {
+            return elItems.filter(item => item.getAttribute('theme') !== theme).length === 0;
+          }
+
+          return el.getAttribute('theme') === theme;
+        }
+
+        ['propagate', 'remove'].forEach(condition => {
```
Comment (tomivirkki) [inline test/items.html]: This iterator only makes the test suite messy. Would be much easier to follow if these were separate tests. I think you only need 2-3 anyway (see the above comment).

File: src/vaadin-contextmenu-items-mixin.html (hunk 2 above)
Comment (yuriy-fix) [reply src/vaadin-contextmenu-items-mixin.html]: Updated.

File: src/vaadin-context-menu-overlay.html (hunk 3 above)
Comment (yuriy-fix) [reply src/vaadin-context-menu-overlay.html]: Replaced with closing the overlay & emptying the root element

File: src/vaadin-contextmenu-items-mixin.html (hunk 4 above)
Comment (yuriy-fix) [reply src/vaadin-contextmenu-items-mixin.html]: Updated.

File: test/items.html (hunk 5 above)
Comment (yuriy-fix) [reply test/items.html]: Removed and tests are simplified. Now one test is testing all three `context-menus` and their nested elements.

File: test/items.html (hunk 6 above)
Comment (yuriy-fix) [reply test/items.html]: Updated. Now there are only 2 tests.

**Comment by yuriy-fix** [comment]
The side-effect is now fixed.

**Comment by yuriy-fix** [comment]
Fixed and test for removal of the `theme` attribute from nested elements is added.

**Pull request merged**